File: src/agents/writer_agent.py
"""
SmartExamBuilder – Writer Agent

Ders kazanımı ve RAG bağlamını alarak GPT-4o-mini ile
açık uçlu sınav sorusu ve puanlama rubriği üretir.
"""

import os
import json
from langchain_openai import ChatOpenAI
from langchain_core.messages import SystemMessage, HumanMessage

from src.schemas.exam_schema import (
    Question, QuestionType, DifficultyLevel, RubricCriteria
)
from src.rag.retriever import retrieve_context


def _extract_json(text: str) -> dict:
    """LLM çıktısından JSON bloğunu güvenli şekilde çıkarır."""
    text = text.strip()
    if "```" in text:
        parts = text.split("```")
        for i, part in enumerate(parts):
            if i % 2 == 1:
                content = part.lstrip("json").strip()
                try:
                    return json.loads(content)
                except json.JSONDecodeError:
                    continue
    return json.loads(text)


def _build_system_prompt(outcome_id: int, difficulty: str) -> str:
    return (
        "Sen deneyimli bir eğitim uzmanısın. Verilen ders kazanımına ve bağlamına göre "
        "yüksek kaliteli, Türkçe açık uçlu bir sınav sorusu ve detaylı puanlama rubriği üretiyorsun.\n\n"
        "SADECE aşağıdaki JSON yapısında yanıt ver, başka hiçbir şey yazma:\n\n"
        "{\n"
        '  "question_text": "Soru metni buraya (Türkçe, açık ve anlaşılır)",\n'
        '  "question_type": "acik_uclu",\n'
        f'  "learning_outcome_id": {outcome_id},\n'
        f'  "difficulty": "{difficulty}",\n'
        '  "expected_keywords": ["anahtar1", "anahtar2", "anahtar3", "anahtar4"],\n'
        '  "rubric": {\n'
        '    "excellent": "Mükemmel cevabın ne içermesi gerektiğinin açıklaması (90-100 puan)",\n'
        '    "good": "İyi cevabın ne içermesi gerektiğinin açıklaması (70-89 puan)",\n'
        '    "satisfactory": "Yeterli cevabın ne içermesi gerektiğinin açıklaması (50-69 puan)",\n'
        '    "insufficient": "Yetersiz cevabın tanımı (0-49 puan)",\n'
        '    "max_points": 100\n'
        "  }\n"
        "}\n\n"
        "Önemli kurallar:\n"
        "- Soru doğrudan kazanımı ölçmeli\n"
        "- Rubrik somut ve ölçülebilir kriterler içermeli\n"
        "- Beklenen anahtar kelimeler teknik ve anlamlı olmalı\n"
        "- Zorluk seviyesine uygun bir derinlik kullan"
    )


def generate_question(
    learning_outcome: str,
    outcome_id: int,
    difficulty: str = "orta"
) -> Question:
    """
    Verilen kazanım için Writer Agent'ı kullanarak soru üretir.

    Args:
        learning_outcome: Ders kazanımı metni
        outcome_id:       Kazanım numarası (1-4)
        difficulty:       Zorluk seviyesi ('kolay', 'orta', 'zor')

    Returns:
        Oluşturulan Question nesnesi
    """
    llm = ChatOpenAI(
        model="local-model", # LM Studio'da yüklü olan modeli kullanır
        temperature=0.7,
        base_url="http://localhost:1234/v1",
        api_key="lm-studio"
    )

    # RAG'dan ilgili bağlamı çek
    context = retrieve_context(learning_outcome, k=3)

    system_msg = _build_system_prompt(outcome_id, difficulty)
    human_msg = (
        f"Ders Bağlamı (RAG'dan çekildi):\n{context}\n\n"
        f"Kazanım: {learning_outcome}\n"
        f"Zorluk: {difficulty}\n\n"
        "Bu kazanıma uygun bir soru ve rubrik üret."
    )

    messages = [
        SystemMessage(content=system_msg),
        HumanMessage(content=human_msg),
    ]

    response = llm.invoke(messages)
    data = _extract_json(response.content)

    # Enum alanlarını güvenli şekilde dönüştür (küçük modeller geçersiz değer üretebilir)
    try:
        q_type = QuestionType(data.get("question_type", "acik_uclu"))
    except ValueError:
        q_type = QuestionType.OPEN_ENDED

    try:
        diff_level = DifficultyLevel(data.get("difficulty", difficulty))
    except ValueError:
        diff_level = DifficultyLevel(difficulty)

    return Question(
        question_text=data["question_text"],
        question_type=q_type,
        learning_outcome_id=data.get("learning_outcome_id", outcome_id),
        difficulty=diff_level,
        expected_keywords=data.get("expected_keywords", []),
        rubric=RubricCriteria(**data["rubric"]),
    )

File: backend/services/writerAgent.js
/**
 * SmartExamBuilder – Writer Agent
 *
 * Python writer_agent.py mantığının JavaScript karşılığı.
 * Ders kazanımı ve RAG bağlamını kullanarak açık uçlu sınav sorusu üretir.
 */

const { callLLM, extractJSON } = require("./llmService");
const { retrieveContext } = require("./ragService");

/**
 * Writer Agent için sistem promptu oluşturur.
 */
function buildSystemPrompt(outcomeId, difficulty) {
  return (
    "Sen deneyimli bir eğitim uzmanısın. Verilen ders kazanımına ve bağlamına göre " +
    "yüksek kaliteli, Türkçe açık uçlu bir sınav sorusu ve detaylı puanlama rubriği üretiyorsun.\n\n" +
    "DİKKAT: JSON formatındaki örnek değerleri AYNEN YAZMA! Lütfen üretilmiş kendi gerçek sorunu ve rubriğini yaz.\n" +
    "SADECE aşağıdaki JSON yapısında yanıt ver, başka hiçbir şey yazma:\n\n" +
    "{\n" +
    '  "question_text": "<ÜRETTİĞİNİZ_GERÇEK_SORU_METNİ>",\n' +
    '  "question_type": "acik_uclu",\n' +
    `  "learning_outcome_id": ${outcomeId},\n` +
    `  "difficulty": "${difficulty}",\n` +
    '  "expected_keywords": ["<anahtar_1>", "<anahtar_2>"],\n' +
    '  "rubric": {\n' +
    '    "excellent": "<90-100 puanlık çok iyi cevabın içeriği>",\n' +
    '    "good": "<70-89 puanlık iyi cevabın içeriği>",\n' +
    '    "satisfactory": "<50-69 puanlık yeterli cevabın içeriği>",\n' +
    '    "insufficient": "<0-49 puanlık yetersiz cevabın içeriği>",\n' +
    '    "max_points": 100\n' +
    "  }\n" +
    "}\n\n" +
    "Önemli kurallar:\n" +
    "- Soru doğrudan kazanımı ölçmeli.\n" +
    "- Rubrik somut ve ölçülebilir kriterler içermeli.\n" +
    "- Beklenen anahtar kelimeler teknik ve anlamlı olmalı.\n" +
    "- Geri dönüş sadece geçerli bir JSON objesi olmalı."
  );
}

/**
 * Verilen kazanım için soru üretir.
 *
 * @param {string} learningOutcome - Kazanım metni
 * @param {number} outcomeId       - Kazanım numarası (1-14)
 * @param {string} difficulty      - Zorluk seviyesi ('kolay', 'orta', 'zor')
 * @returns {Promise<object>}      - Üretilen soru objesi
 */
async function generateQuestion(learningOutcome, outcomeId, difficulty = "orta") {
  // RAG bağlamını al
  const context = retrieveContext(outcomeId);

  const systemPrompt = buildSystemPrompt(outcomeId, difficulty);
  const userPrompt =
    `Ders Bağlamı (RAG'dan çekildi):\n${context}\n\n` +
    `Kazanım: ${learningOutcome}\n` +
    `Zorluk: ${difficulty}\n\n` +
    "Bu kazanıma uygun bir soru ve rubrik üret.";

  const response = await callLLM(systemPrompt, userPrompt, 0.7);
  const data = extractJSON(response);

  // Geçerli enum değerlerini kontrol et
  const validTypes = ["acik_uclu", "coktan_secmeli"];
  const validDifficulties = ["kolay", "orta", "zor"];

  return {
    question_text: data.question_text,
    question_type: validTypes.includes(data.question_type) ? data.question_type : "acik_uclu",
    learning_outcome_id: data.learning_outcome_id || outcomeId,
    difficulty: validDifficulties.includes(data.difficulty) ? data.difficulty : difficulty,
    expected_keywords: data.expected_keywords || [],
    rubric: {
      excellent: data.rubric?.excellent || "",
      good: data.rubric?.good || "",
      satisfactory: data.rubric?.satisfactory || "",
      insufficient: data.rubric?.insufficient || "",
      max_points: data.rubric?.max_points || 100,
    },
  };
}

module.exports = { generateQuestion };
